Add ConfigDocument API

fpringvaldsen opened the pull request:
This work came out of the discussion in https://github.com/typesafehub/config/issues/272.

~~These commits add a new grouping of classes, `ConfigNode`, to the library, which are used to form an AST. This AST contains all the tokens from which it was parsed, so the original text can be reproduced from them.~~

~~I'm opening this up for review now as the implementation is complete, so I think that portion is ready for review. The testing code is rough right now, largely because of the difficulty involved with crafting instances of `ConfigNodeComplexValue` by hand, so I'm still working on improving that, although any feedback on that front is welcome.~~

This PR adds the ConfigDocument API in its entirety, including its backend representation with ConfigNodes, the ConfigDocument Parser, and the ConfigDocument Interface.

havocp commented on config/src/main/java/com/typesafe/config/impl/AbstractConfigNode.java:
```diff
@@ -0,0 +1,15 @@
+import com.typesafe.config.ConfigNode;
+
+abstract class AbstractConfigNode implements ConfigNode {
+    final private Token token;
```
it seems like the "responsibility" of this class is to be a node with only one token, so `ConfigNodeSingleToken` or something?

Some nodes will have multiple tokens right? Or not? I would expect the base class for all nodes to abstract that a little, maybe with something like `protected Iterable tokens()` and then for `ConfigNodeSingleToken` that would return `Collections.singletonList(token)` or the like.

havocp replied:
then `render` would of course render the whole list of tokens. but maybe we have no multi-token nodes?

fpringvaldsen replied:
@havocp As it stands right now, each token lives in its own node. However, some nodes have child nodes instead of Tokens.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.*;
+
+final class ConfigNodeComplexValue implements ConfigNode, ConfigNodeValue {
+    private ArrayList<ConfigNode> children;
```
should this be final? it doesn't change after the constructor right?

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,95 @@
+
+final class ConfigNodeComplexValue implements ConfigNode, ConfigNodeValue {
+    private ArrayList<ConfigNode> children;
+    final private LinkedHashMap<Path, Integer> map = new LinkedHashMap<>();
```
I would expect the diamond `<>` notation to fail with `-source 1.6` that we are using right now, but we could probably change that to `-source 1.7`

fpringvaldsen replied:
@havocp Whoops, this map wasn't actually supposed to be here. I'll remove it, so this shouldn't be a problem.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,95 @@
+            if (currNode instanceof ConfigNodeKeyValue) {
+                keyValueIndexes.add(i);
+            }
+        }
```
I'd probably say skip this; it's precomputing an index that's only used in `changeValueOnPath` (correct?) but for most nodes we will never change a value, so we're permanently using memory and doing CPU work up-front that often won't be necessary. We could compute it on the fly only when we actually change a value.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,95 @@
+        }
+    }
+
+    public ArrayList<ConfigNode> children() {
```
I'd return something more general here like `List`, `Collection`, or `Iterable`

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,95 @@
+        for (ConfigNode child : children) {
+            renderedText.append(child.render());
+        }
+        return renderedText.toString();
```
I would expect that there are some tokens "outside" of the children (the `[]` for a list for example?)

also, maybe `render()` that returns a String could be final in `AbstractConfigNode`, and then have a protected method `protected void render(StringBuilder builder)`, so `render()` would chain to the protected method. That way to render the entire document from the root, there's only one StringBuilder.

havocp replied:
oh, but if render() always just means "render all the tokens", this can be even simpler: have a final render method that calls `tokens` and the subtypes only have to implement `tokens()`

fpringvaldsen replied:
@havocp I really like this idea, and have an implementation in place locally. However, I've run into an issue.

As the code currently stands in this PR, only `ConfigNodeBasic`, `ConfigNodeKey,`, and `ConfigNodeSimpleValue` subclass `AbstractConfigNode`, as it's built to be a single-token node. ALL nodes implement the `ConfigNode` interface, and `ConfigNodeSimpleValue` and `ConfigNodeComplexValue` both implement the `ConfigNodeValue` interface, which extends the `ConfigNode` interface. To get this implementation, I definitely think it's best to make all node types subclass `AbstractConfigNode`, which I've done, since the tokens are an internal implementation detail and so the `tokens()` method should not be public-facing. 

However, this is leading to some casting issues. I've changed all nodes with children to keep an `ArrayList` of `AbstractConfigNode`s since they're going to need to produce the tokens of their children using the `tokens` method. However, I still need some way to differentiate the value node types from the other node types. `ConfigNodeValue` can't extend `AbstractConfigNode` since it's an interface, and thus, a `ConfigNodeValue` cannot be cast to an `AbstractConfigNode`, which means they can't be added to the list of children on a node. I could keep the list of children as a list of `ConfigNode`s instead of `AbstractConfigNode`s and have `ConfigNodeValue` extend `ConfigNode`, but then every single node is going to need to be casted when the `tokens()` function is called.

My solution to this is to have an abstract class called `AbstractConfigNodeValue`. This class does nothing and has no special variables, but subclasses `AbstractConfigNode`, which allows type differentiation. However, I'm feeling pretty iffy on this, since it requires the existence of a class that doesn't do anything except allow for type differentiation. I run into the same issue with the `ConfigNodeValue` interface, since that extends `ConfigNode` and doesn't do anything differently from `ConfigNode`. 

@havocp or @cprice404 do either of you have any suggestions on how to get around this problem? It seems like we need to differentiate the value nodes for parameter passing when setting values in `ConfigNodeKeyValue` and `ConfigNodeComplexValue`, but I can't think of another way to differentiate them apart from this, especially since `ConfigNodeComplexValue` and `ConfigNodeSimpleValue` have almost nothing in common apart from the fact that they are both nodes.

havocp replied:
I think this is essentially the same issue as the `ConfigValue` hierarchy; there we have `AbstractConfigValue` which is extended by all the concrete value types, and we usually use `AbstractConfigValue` internally because it can have impl-private methods on it. We also have `AbstractConfigObject extends AbstractConfigValue`.

There's a good chance I bet that your `AbstractConfigNodeValue` would end up having a method on it eventually, but if it doesn't it isn't the end of the world.

If you have some casts to start with we can always clean it up later, I wouldn't get too stuck on it for the moment. Things may become clearer as you get more of the PR done.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    protected ConfigNodeComplexValue changeValueOnPath(Path desiredPath, ConfigNodeValue value) {
+        ArrayList<ConfigNode> childrenCopy = (ArrayList<ConfigNode>)(children.clone());
```
you can copy a list with `new ArrayList<ConfigNode>(children)` and avoid a cast

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,95 @@
+            ArrayList<ConfigNode> childrenCopy = (ArrayList<ConfigNode>)children.clone();
+            ArrayList<ConfigNode> newNodes = new ArrayList<ConfigNode>();
+            newNodes.add(new ConfigNodeBasic(Tokens.newLine(null)));
+            newNodes.add(new ConfigNodeKey(Tokens.newUnquotedText(null, desiredPath.render())));
```
Paths are more complicated than just unquoted text; they can be something like `foo."bar+@#$".baz` or whatever. See `parsePathExpression` in Parser.java. So I think we'll need a multi-token node representing a path, and we'll need to implement a path-to-tokens function.

fpringvaldsen replied:
It should be pretty easy to take the `ConfigNodeKey` class and make it multi-token, so I'll get to work on that. I think we can easily keep track of the tokens used to parse a path in a Collection, then pass that into the `PathBuilder` so that it saves them and can thus reproduce the original tokens/text.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeKeyValue.java:
```diff
@@ -0,0 +1,48 @@
+    final private ArrayList<ConfigNode> children;
+    private int configNodeValueIndex;
+    private ConfigNodeKey key;
+    private ConfigNodeValue value;
```
since these three fields are all caches (they could be computed on the fly), I'd default to computing them on the fly, and only cache if we have to.

fpringvaldsen replied:
Hmm, sounds good. I'm guessing that should be fine, since in most cases I think a key and a value will probably only be checked once, and a key-value node will probably not contain a large number of children.

havocp commented:
To avoid crafting complex nodes by hand, one approach would be to go ahead and implement the tokens-to-ConfigNode parser. That would also allow implementing a round-trip test that no tokens or formatting are lost in the parsing process and that the `ConfigNode` type hierarchy is complete. (Round trip test means:  `configNodeParser("any valid config stuff here").render().equals("any valid config stuff here")` ... this can be tested for the files that are already in src/test/resources perhaps)

I think you could cut-and-paste Parser.java; the changes I see the need to make right away are:

 * you can drop all that TokenWithComments / attractsLeading/TrailingComments stuff and instead just generate comments nodes
 * generate nodes for, rather than discarding, tokens such as ignored whitespace and field separators
 * don't actually process includes, just make nodes for them
 * might want to split the path expression parsing stuff out of Parser.java into its own file so you can use it too

I think if your new parser doesn't have the TokenWithComments stuff and shares the path expression parsing stuff, it won't be that many lines of code, and the lines it has will be based on the Parser.java lines. Most of the verbosity in Parser.java is about nice error messages (which would indeed be nice to keep). Don't be afraid of cut-and-paste for the bulk of the parser, since we can clean that up later by making the existing Parser.java go from `ConfigNode` => `ConfigValue` instead of using tokens.

fpringvaldsen commented:
@havocp Yeah, I was thinking that implementing the parser would make testing easier, but I felt like there was still some value in directly crafting config nodes for tests. 

I have some more tests of config nodes that I haven't pushed up yet. Do you want me to push those up? And would you prefer the Parser be added to this PR or a new PR? I was planning to add in the parser in a new PR after this was merged, but I can add it into this PR if you'd like.

fpringvaldsen commented:
@havocp I've added some more tests and will get to work on addressing your feedback, thanks!

havocp commented:
I agree that some direct tests for the nodes are good too. I'm thinking it might be worth adding the parser here because I'm having trouble feeling we have the node API right without also having the parser and roundtrip tests.  What do you think?

What I would like to keep separate is the modification of Parser.java - mostly because that's when we risk breaking existing functionality. It would be nice if this PR is only adding new stuff, and then make it a separate step to unify with the old stuff.

(exception: maybe that path expression parser code can move into its own PathParser file, since that's hopefully just a simple move.)

fpringvaldsen commented:
@havocp Hmm, that makes sense to me. This feature isn't really finished without the Parser, so if they get merged separately, we'll have an incomplete, in-progress feature in the repository. I'll address your feedback, then get to work on the new Parser. I'll add it as a commit to this PR once it's ready to be reviewed.

fpringvaldsen commented:
@havocp I've just pushed up a new commit addressing most of your feedback. In regards to the path changes, I'm getting to work on that now, and will push that up in a separate commit.

fpringvaldsen commented:
@havocp I've pushed up a new commit that saves the Tokens in Paths produced by `Parser.parsePath`. It should be ready for review, so any feedback is much appreciated!

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,89 @@
+        return children;
+    }
+
+    protected Collection<Token> tokens() {
```
this one should have `@Override`

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,89 @@
+        this.children = new ArrayList(children);
+    }
+
+    public Iterable<AbstractConfigNode> children() {
```
this can be final I guess

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,89 @@
+        ArrayList<AbstractConfigNode> childrenCopy = new ArrayList(children);
+        boolean replaced = value == null;
+        ConfigNodeKeyValue node;
+        Path key;
```
these two vars could be inside the loop right?

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeKey.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.Collection;
+
+final class ConfigNodeKey extends AbstractConfigNode {
```
there's a convention throughout the lib, or at least in public API, that we use "key" to mean one element of a path; paths have to be parsed, keys are the raw strings in each Path element. ConfigNodeKey should probably be ConfigNodePath to be sure we don't get confused on this point.

fpringvaldsen replied:
Sounds good. I think it would also be worth changing `ConfigNodeKeyValue` to `ConfigNodePathValue` so I'll go ahead and do that too.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,89 @@
+            node = (ConfigNodeKeyValue)children.get(i);
+            key = node.key().value();
+            if (key.equals(desiredPath)) {
+                if (!replaced) {
```
it's a nice practice to avoid combining negation and if/else (do `if (replaced) { remove(i) } else {}`)

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -0,0 +1,89 @@
+                }
+                else
+                    childrenCopy.remove(i);
+            } else if (desiredPath.startsWith(key)) {
```
below feels like it should be using recursion somehow - maybe it's as simple as `childrenCopy.set(i, node.changeValueOnPath(remainingPath, value))` ?

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeKeyValue.java:
```diff
@@ -0,0 +1,55 @@
+        throw new ConfigException.BugOrBroken("KeyValue node doesn't have a value");
+    }
+
+    public ConfigNodeKey key() {
```
this should probably be called path() and the KeyValue class could either be called `PathValue` or perhaps simply `Field` would be easier.

havocp commented on config/src/main/java/com/typesafe/config/impl/Parser.java:
```diff
@@ -1063,6 +1063,7 @@ private static Path parsePathExpression(Iterator<Token> expression,
     private static Path parsePathExpression(Iterator<Token> expression,
             ConfigOrigin origin, String originalText) {
         // each builder in "buf" is an element in the path.
+        ArrayList<Token> pathTokens = new ArrayList<Token>();
```
It feels like a somewhat messy mixing of concerns to make `Path` and `PathBuilder` keep the tokens around; we do use `Path` in a performance-sensitive context when we parse paths passed to the getters on `Config`, like `Config.getInt`. That's why there's all this fast-path path-parsing stuff. But, even ignoring performance I think it's a little messy. What I'd suggest is to factor this code so there's a version that returns a `Path` and a version that returns a `ConfigNodeKey`, and then a private function which maybe takes the `pathTokens` array list which can be null. So:

```
private static Path parsePathExpression(Iterator<Token> expression, ConfigOrigin origin, String originalText, ArrayList<Token> pathTokens) {
   ...
   if (pathTokens != null) pathTokens.add(t)
   ...
}
```

Then you can have the functions which call that either provide the `ArrayList` or provide null, depending on whether they want to end up with a `Path` or `ConfigNodeKey`

It would be fine to skip the `speculativeFastParsePath` codepath for `ConfigNodeKey`, and always use the slow `parsePathExpression` when we need the tokens. The fast path matters only for that `Config.getInt` case.

fpringvaldsen replied:
I think that's doable, but I think in that case, a `ConfigNodePath` is going to need to be passed into the `setValueOnPath()` method in `ConfigNodeComplexValue`, since we'll need the tokens to add in a new field.

havocp commented on config/src/main/java/com/typesafe/config/impl/Path.java:
```diff
@@ -141,6 +154,21 @@ Path subPath(int firstIndex, int lastIndex) {
+    boolean startsWith(Path other) {
+        Path myRemainder = this;
+        Path otherRemainder = other;
+        if (otherRemainder.length() <= myRemainder.length()) {
```
since length is O(n) it would be a bit nicer to instead check myRemainder for null, below, perhaps. Since paths are generally short it probably doesn't really matter that much.

havocp commented:
Looking better all the time, I appreciate the work on this. I think it's turning out to be fairly simple, maybe more so than I expected... not a common experience in software development :-) yay!

cprice404 commented:
@havocp we owe you a beer for all your help on this! :)

fpringvaldsen commented:
@havocp I've addressed your comments, and have also extracted the Path Parsing into a new class. Let me know what you think!

fpringvaldsen commented:
@havocp Quick question: for the new Parser, should I be keeping the logic that throws errors if the parsed text does not conform to JSON? Or should we ignore that in the `ConfigDocumentParser`, and let the current Parser handle it when it grabs config nodes from the `ConfigDocumentParser`?

havocp commented:
I think when loading a document you might want to require it to be JSON-conforming, so maybe the doc parser should be able to check JSON conformance too.

fpringvaldsen commented:
@havocp @cprice404 I've just pushed up my implementation of the parser if either of you want to take a look. It can parse a file into a ComplexConfigNode and render that file back out exactly as it was rendered in.

I don't have a `ConfigDocument` class implemented yet, that's coming next. Once that's implemented, I'm going to add `ConfigDocument` tests that test replacement and rendering (since `ConfigNodeComplexValue` does the lion's share of the work, I doubt a `ConfigDocument` implementation will take me long at all).

I've put this up now as I feel it's ready for review. Hopefully I'll have the `ConfigDocument` changes up by some point tomorrow, at which point this PR will have all it needs for final review/merge.

fpringvaldsen commented on config/src/main/java/com/typesafe/config/impl/ConfigDocumentParser.java:
```diff
@@ -0,0 +1,620 @@
+                    // there may be newlines _within_ the objects and arrays
+                    v = parseValue(t);
+                    valueCount++;
+                } else if (Tokens.isIgnoredWhitespace(t)) {
```
I'm planning to refactor this a bit before I work on the `ConfigDocument` stuff, since if it sees a complex value like an object or an array, if there is no concatenation, it will throw out the value it parsed, which could be a large performance hit in the cases of very large objects or arrays.

fpringvaldsen commented:
@havocp The Parser is finished, and is now ready for review. Any feedback (especially on test coverage) is welcome!

havocp commented:
fantastic! I'm at scala days right now and iPhone-only most of the day, so my review may be lagged as late as next week, but I will try to take a first pass this evening or throw in a few impressions via iPhone. 

One thing you can do is run jacoco:cover (do "project config" first I think to enter the config subproject). then look at the generated html report under config/target.

I haven't looked at the code yet but another idea would be to share all applicable tests with the old parser, for example they should probably throw most of the same errors.

you can also run findbugs which may turn up a couple small issues, or not. sometimes it does.

havocp commented:
Were there any interesting changes or decisions porting the old parser to generate nodes?

fpringvaldsen commented:
Thanks! I'll look into those approaches, probably once I get the `ConfigDocument` stuff pushed up. 

There weren't a ton of super noteworthy changes, it was pretty easy to modify the Parser to create nodes. Mostly I was able to simplify things quite a bit and cut out a number of methods, which was nice, it was a pretty fun exercise.

fpringvaldsen commented:
Okay, got the `ConfigDocument` stuff written along with some tests. This is ready for final review, I think. 

The only problem I know of now is adding a new path in JSON files. This is tricky because of the way JSON works, as all keys need to be quoted. What's happening with CONF files is the path is just being added at the top level, so if you give it path `a.b.c`, and it doesn't see that path anywhere in the file, it'll just add `a.b.c=[value]` to the top level. This is not valid in JSON, as all paths need to be quoted, and we can't just add a quote around the path when adding a new path to a JSON input because if we were replacing a value with a multi-element path, we'd add path `"a.b.c"` instead of `a.b.c`, and those are different.

I think the path replacement is going to have to be modified somehow. I can think of three approaches:

* We don't support adding a new path in JSON if it's not a single-element path.
* We support adding all new paths in JSON, but when we add a multi-element path, we quote the entire thing, turning it into a single-element path.
* Instead of adding the path at the top level, we move down through the tree until we find the object where only a single-element path remains, and add the value into that object with the necessary key. I definitely think this is the best approach. So if we had the map
```
{a: b
 c: {
   d: e
 }
```
And set value `12` at path `c.f`, we'd get the following map:

```
{a: b
  c: {
    d: e
    f: 12
  }
```

In this case, if there does not exist an object wherein the remaining path is single-element, then we'd add the full path to the top level of the map in the case of parsing a CONF file, or throw an exception in the case of a JSON file. So, if in the above example, we wanted to set value `12` at path `c.f.g`, an exception would be thrown if the file were JSON, but we'd get the following if it were a CONF file:

```
{a: b
  c: {
    d: e
  c.f.g: 12  
}
```

I'm thinking the third approach would be best, especially now that I've put it down into writing and can see that it probably won't be very hard to accomplish at all. 

@havocp How does this sound to you? Would you prefer a different approach?

havocp commented:
I like the third approach. To be clear, we would throw if we try to set a field in a non-object right, but if we try to set a field in a path that doesn't exist we could auto create the parent object I would think?

havocp commented on config/src/main/java/com/typesafe/config/ConfigDocument.java:
```diff
@@ -0,0 +1,38 @@
+     * at the final occurrence of the path. If the path does not exist, it will be added.
+     *
+     * @param path the path at which to set the desired value
+     * @param newValue the value to set at the desired path
```
This should be very clear about what the string is - I guess it's the rendered value? so we might call the paramter newValueText and document that "this text will be placed in the file as-is with no additional parsing or escaping"? or is that not right ? I haven't gotten down to the imementation yet :-)

fpringvaldsen replied:
It's the value as you would want it to appear in the config file. The implementation actually uses the `ConfigDocumentParser` to parse the value string and generate the node which it will add into the document.

havocp commented on config/src/main/java/com/typesafe/config/ConfigDocumentFactory.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Factory for automatically creating a ConfigDocument from a given input. Currently
+ * only supports files and strings.
+ */
```
we might want Reader too I suppose. Then people can plug in other kinds of stream.

fpringvaldsen replied:
Alright, I'll add support for a Reader as well.

havocp commented:
so right now the public ConfigNode is useless right (there's no way to get one?)

That's fine and even good (would be nice to get this basic document with set value thing done, then look at public nodes later, I don't know if you even want that).
But wanted to clarify.

fpringvaldsen commented:
In terms of the value setting, yeah, I think we could definitely create the parent object if it doesn't exist. I think it will be tricky, though, to figure out which level on which to create the parent object and the indentation for the parent object/new setting.

And yeah, right now there's no way to use the public ConfigNode. I'm thinking we can add that in a subsequent PR.

cprice404 commented:
I would think maybe it'd be alright to not worry about indentation for the first pass at this?

fpringvaldsen commented:
@cprice404 That's a good point, that can probably be added in a subsequent PR. If we're just worrying about adding the parent object, this should be quite a bit easier to implement. 

I'll work on modifying the value adding so it creates the parent object if it doesn't exist. I think that'll be pretty easy if we're not worrying about indentation, so hopefully I'll have something ready in the next couple hours.

havocp commented:
ignoring indentation seems fine to me as a first cut, go for it.

fpringvaldsen commented:
Okay, I've fixed up the field addition. Now, all non-existent objects along the path will be created, and when the `ConfigDocument` was parsed from a JSON document, the resultant `ConfigDocument` will render valid JSON.

fpringvaldsen commented:
I've updated the title, since this is now in a state where I feel it is ready for a full review with the intent of being merged.

fpringvaldsen commented on config/src/test/scala/com/typesafe/config/impl/ConfigDocumentParserTest.scala:
```diff
@@ -0,0 +1,258 @@
+  @Test
+  def parseJSONFailures() {
+      // JSON does not support concatenations
+      parseJSONFailuresTest("""{ "foo": 123 456 789 } """, "Expecting close brace } or a comma")
```
All of these exceptions yield the same messages as the ones being thrown by the current parser, with the exception of this one. However, I'm thinking this error message is somewhat more informative (the error message yielded by the other parser states that `' '` is an invalid token in JSON)

havocp replied:
Agreed, the new message seems better.

fpringvaldsen commented:
I'm currently working on a new PR on top of this current one to refactor the current parser to use the new parser. Some internal implementation details will be changed, but the results of the public API as it stands in this PR will not be changing, so this should still be good for merging.

havocp commented:
apologies for being slow here, have been traveling and so on.

I'm thinking there might be some work to make the ConfigNode subtypes a little more "semantic," like have nodes for includes and comments and so on. I'd think the public API has stuff like ConfigNodeComment rather than ConfigNodeSingleToken. The old parser then would do very little (drop purely syntactic nodes, process includes, ...)

Is that part of what you're doing now or does it not make sense ?

fpringvaldsen commented:
@havocp Yeah, that's part of what I'm doing now for the subsequent PR. If you'd like I can merge what I've done (minus the Parser changes) into this PR.

fpringvaldsen commented:
Also, no problem on the response time! We know you've been busy with Scala Days and such. :)

havocp commented:
It's fine to keep it separate; I'll get this one merged. I don't think I have any more substantive comments on this PR I just wanted to at least read all the code.

fpringvaldsen commented:
@havocp Great! Let me know if you have any feedback, I'm happy to make any more changes you think are necessary (particularly in regards to test coverage).

fpringvaldsen commented:
@havocp I noticed you had given me some feedback around the `ConfigDocument` that I hadn't addressed yet, I pushed up a new commit that addresses it.

havocp commented on config/src/main/java/com/typesafe/config/ConfigDocument.java:
```diff
@@ -0,0 +1,40 @@
+     * @param path the path at which to set the desired value
+     * @param newValue the value to set at the desired path, represented as a string. This
+     *                 string will be parsed into a ConfigNode, and the text will be inserted
+     *                 as-is into the document, with leading and trailing whitespace removed.
```
trimming whitespace surprises me a little - if we're going to some lengths here to let people specify the exact text they want, why not let them have extra whitespace? they can trim it if they want ...

havocp replied:
if you parse a JSON file then set a value that uses HOCON syntax, what happens?

fpringvaldsen replied:
In terms of whitespace trimming, this was primarily due to the fact that the desired value could be a simple value, which does not support multiple tokens/nodes. We could modify that for sure, though.

In terms of your question on value-setting: the ConfigDocument object saves the options which were used to parse it, and then uses those to parse any values passed to the `setValue()` method. If you parse a JSON file then set a value that uses HOCON syntax, it should conceivably throw a parse error. I don't have a test for that, but could easily add one.

havocp replied:
seems worth testing to me, since breaking that behavior could certainly break an app (and it seems easy to break)

fpringvaldsen replied:
I've written a test and verified the behavior. 

One thing to note: when parsing a single JSON value for node replacement, if a concatenation is given, the parser will simply parse the first value in the concatenation into a node and ignore all other items in the concatenation. Is that acceptable? Or would you prefer it error out when a concatenation is given and JSON is being parsed?

Also, do you want me to modify the `setValue` method to NOT trim whitespace? I think it could get a little weird, since this is written in such a way that the parent object keeps track of leading and trailing whitespace, so we'd have to either stick the value with leading and trailing whitespace into a concatenation node, make a new wrapper for simple config nodes that contains the whitespace, make `ConfigNodeSimpleValue` multi-token but not multi-node, or have a way to insert multiple nodes at once into a config node instead of just one.

fpringvaldsen replied:
Actually, I'm thinking that a concatenation is a non-valid value in JSON, so I'll have it throw an exception if there's a concatenation in JSON.

havocp replied:
I think it should probably parse the entire concatenation, which would be a "value" in this sense, but if that's hard it would be better to throw an error than to silently ignore stuff.

I guess if we don't trim whitespace a weird thing results where setting a value isn't idempotent, unless we made setting a value remove all whitespace around the existing value. Not desirable. What about comments? they pose similar issues.

A simple punt would be "anything except the value itself is an error" - that would allow us to be more forgiving in the future without breaking people.

havocp replied:
oh I missed that this was in JSON. agreed invalid json should throw.

fpringvaldsen replied:
Hmm, I'm not entirely sure how parsing the whole concatenation would work if it's JSON. Since the rendered output of the node created by the single value parsing needs to be valid JSON, we'd need to modify the rendered result so it's valid JSON. If we're combining simple values that's not a huge deal (can just put quotes around the rendered output) but when we get to objects and arrays we run into a problem wherein we need to merge not just the values but the text contained within them to form a single object, which seems pretty complicated.

Sounds like throwing an error if there's surrounding whitespace might be a good way to go. I could also add a new Node type or make ConfigNodeSimpleValue multi-token, though, either of those two approaches shouldn't be difficult.

havocp commented on config/src/main/java/com/typesafe/config/ConfigDocument.java:
```diff
@@ -0,0 +1,40 @@
+     *                 as-is into the document, with leading and trailing whitespace removed.
+     * @return a copy of the ConfigDocument with the desired value at the desired path
+     */
+    ConfigDocument setValue(String path, String newValue);
```
it can go in a later PR but a natural method to add here would be one that takes a ConfigValue rather than a String (presumably it would just chain over to the string one by doing `value.render()` though)

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java:
```diff
@@ -0,0 +1,127 @@
+                if (node.value() instanceof ConfigNodeObject) {
+                    Path remainingPath = desiredPath.subPath(key.length());
+                    childrenCopy.set(i, node.replaceValue(((ConfigNodeObject)node.value()).changeValueOnPath(remainingPath, valueCopy)));
+                    if (valueCopy != null && node.render() != super.children.get(i).render())
```
I guess the `!=` here should be `equals`, but I would think if the renders are equal the original nodes are also equal ... if you want to implement equals() with render(), you could do that in AbstractConfigNode itself perhaps

havocp commented:
The test coverage looks reasonable to me, I'm also thinking that once you port the existing parser to use this one, we'll get a ton of coverage "for free" from all the existing tests.

havocp commented:
the only outright bug I saw was the `!=`/`equals` thing, so unless I'm confused there we should probably fix it. otherwise, I've been through this code and it looks good to merge.

fpringvaldsen commented:
@havocp Great! I'll go through and address your last bits of feedback, hopefully it won't take long

fpringvaldsen commented:
Thanks for the review!

havocp commented:
throwing on any kind of stuff (comments, whitespace, concatenation not supported by JSON) around the value in `setValue` makes sense to me, because otherwise it gets into all kinds of thorny issues where we can have multiple trees possible for the same rendered file.

fpringvaldsen commented:
@havocp Alright, I'll get to work now on making a new simple value class that is multi-node to contain whitespace/comments, which will be used solely in the case of setting a value.

havocp commented:
I was trying to say don't do that - I don't think we want the whitespace/comments to be part of the value node sometimes, and part of the parent node sometimes. Also, I don't think we want setting a value to be non-idempotent. Instead, setValue could give up and throw if there's any "extras" around the value. That way we can punt on all the semantics of whitespace/comments.

fpringvaldsen commented:
Ah, sorry, misunderstood what you were saying! That should be a fairly quick fix.

fpringvaldsen commented:
@havocp Alright, I've addressed your last few comments!

havocp commented on config/src/test/scala/com/typesafe/config/impl/ConfigDocumentTest.scala:
```diff
@@ -0,0 +1,257 @@
+                assertTrue(e.getMessage.contains("ConfigDocument had an array at the root level"))
+        }
+        assertTrue(exceptionThrown)
+    }
```
for future reference, you can write this as `val e = intercept[ConfigException] { document.setValue("a", "1") }` which asserts that the exception of that type is in fact thrown, then you can check the message.

havocp commented:
Merged, thanks very much!
